tmt.spec

```diff
@@ -1,250 +1,109 @@
-BuildRequires: python%{python3_pkgversion}-click
-BuildRequires: python%{python3_pkgversion}-fmf >= 1.2.0
-BuildRequires: python%{python3_pkgversion}-requests
-BuildRequires: python%{python3_pkgversion}-testcloud >= 0.9.2
-BuildRequires: python%{python3_pkgversion}-markdown
-BuildRequires: python%{python3_pkgversion}-junit_xml
-BuildRequires: python%{python3_pkgversion}-ruamel-yaml
-BuildRequires: python%{python3_pkgversion}-jinja2
-# Only needed for rhel-8 (it has python3.6)
-%if 0%{?rhel} == 8
-BuildRequires: python%{python3_pkgversion}-typing-extensions
-BuildRequires: python%{python3_pkgversion}-dataclasses
-BuildRequires: python%{python3_pkgversion}-importlib-metadata
-%endif
-# Required for tests
-BuildRequires: rsync
-%{?python_provide:%python_provide python%{python3_pkgversion}-%{name}}
+BuildArch:      noarch
+BuildRequires:  python3-devel
 
-%description -n python%{python3_pkgversion}-%{name}
+%global _description %{expand:
 The tmt Python module and command line tool implement the test
-metadata specification (L1 and L2) and allows easy test execution.
-This package contains the Python 3 module.
-
-%package provision-container
-Summary: Container provisioner for the Test Management Tool
-Obsoletes: tmt-container < 0.17
-Requires: tmt == %{version}-%{release}
-Requires: podman
-Requires: (ansible or ansible-collection-containers-podman)
-
-%description provision-container
-Dependencies required to run tests in a container environment.
-
-%package provision-virtual
-Summary: Virtual machine provisioner for the Test Management Tool
-Obsoletes: tmt-testcloud < 0.17
-Requires: tmt == %{version}-%{release}
-Requires: python%{python3_pkgversion}-testcloud >= 0.9.2
-Requires: libvirt-daemon-config-network
-Requires: openssh-clients
-Requires: (ansible or ansible-core)
+metadata specification (L1 and L2) and allows easy test execution.}
+
+%description %_description
+
+Summary:        %{summary}
+Requires:       git-core rsync sshpass
+
+Obsoletes:      tmt-all
+Obsoletes:      tmt-report-html
+Obsoletes:      tmt-report-junit
+Obsoletes:      tmt-report-polarion
+Obsoletes:      tmt-report-reportportal
+Obsoletes:      tmt-test-convert
+
+Recommends:     tmt+test-convert
+Recommends:     tmt+export-polarion
+Recommends:     tmt+provision
+Recommends:     tmt+report-junit
+Recommends:     tmt+report-polarion
+
+%description -n tmt %_description
+
+%pyproject_extras_subpkg -n tmt test-convert
+%pyproject_extras_subpkg -n tmt export-polarion
+%pyproject_extras_subpkg -n tmt provision
+%pyproject_extras_subpkg -n tmt report-junit
+%pyproject_extras_subpkg -n tmt report-polarion
+%pyproject_extras_subpkg -n tmt all
+
+%package -n     tmt+provision-container
+Summary:        Dependencies required for tmt container provisioner
+Obsoletes:      tmt-provision-container
+Obsoletes:      tmt-container < 0.17
+Requires:       tmt == %{version}-%{release}
+Requires:       podman
+Requires:       (ansible or ansible-collection-containers-podman)
```

Comment:
When you create an extra subpackage manually like here, you also need to include the `%files` section for it.

Reply:
Even if there are no files? The intention here was that these 'metapackages' should behave same as the other %pyproject_extras_subpkg packages (which don't include non-python rpms dependencies like podman here).

Reply:
Even then, it needs to ghost the dist-info. See here:

```
$ rpm --eval '%pyproject_extras_subpkg -n tmt provision-container'
%package -n tmt+provision-container
Summary: Metapackage for tmt: provision-container extras
Requires: tmt = %{version}-%{release}
%description -n tmt+provision-container
This is a metapackage bringing in provision-container extras requires for tmt.
It makes sure the dependencies are installed.

%files -n tmt+provision-container -f .../%{name}-%{version}-%{release}.x86_64-pyproject-ghost-distinfo



```

Reply:
Manually, that would mean:

```
%files -n tmt+provision-container -f %{_pyproject_ghost_distinfo}
```

(Using a "private" macro -- feel free to open a discussion on the Fedora's Python mailing list about making that macro "public".)

Or:

```
%files -n tmt+provision-container
%ghost %{python3_sitelib}/tmt*.dist-info
```

Using a path.

Reply:
Note that `%package`s without `%files` are not even produced/built. Even if you actually wanted a package without files (a "true" metapackage, so to speak), you would need to have an empty `%files` section for it. But metapackages with Python extras actually have files (but they are `%ghost`ed). It's the implementation detail of RPM dependency generators. Normally, such implementation details are abstracted by using `%pyproject_extras_subpkg` -- but when you want to add custom Obsoletes, Requires, etc., you lose the ability to use that abstraction.

We could add an API to `%pyproject_extras_subpkg` to specify custom Obsoletes, Requires, etc., but I am afraid it could easily grow into a monstrosity :thinking:

Reply:
Thanks, the private macro seems to be working fine. 
"private" means it's not in guidelines?